src/sentry/grouping/strategies/newstyle.py

```diff
@@ -733,6 +740,83 @@ def chained_exception(
+def filter_exceptions_for_exception_groups(
+    exceptions: List[SingleException], context: GroupingContext, event: Event, **meta: Any
+) -> List[SingleException]:
+
+    # This function only filters exceptions if there are at least two exceptions.
+    if len(exceptions) <= 1:
+        return exceptions
+
+    # Reconstruct the tree of exceptions if the required data is present.
+    nodes = {}
+    for exception in reversed(exceptions):
+        exception._children = []
+        mechanism: Mechanism = exception.mechanism
+        if mechanism and mechanism.exception_id is not None:
+            nodes[mechanism.exception_id] = exception
+            if mechanism.parent_id is not None:
+                nodes[mechanism.parent_id]._children.append(exception)
+
+    # If the nodes list is empty, then we don't have the required data to reconstruct the tree.
+    if len(nodes) == 0:
+        return exceptions
+
+    # Traverse the tree recursively from the root node to get all "top-level exceptions"
+    # and sort for consistency.
+    top_level_exceptions = sorted(
+        get_top_level_exceptions(nodes[0]), key=lambda exception: str(exception.type), reverse=True
+    )
+
+    # Figure out the distinct exceptions, grouping by the hashes of their grouping components.
+    distinct_exceptions = [
+        next(group)
+        for _, group in itertools.groupby(
+            top_level_exceptions,
+            key=lambda exception: [
+                component.get_hash()
+                for component in context.get_grouping_component(
+                    exception, event=event, **meta
+                ).values()
+            ],
+        )
+    ]
+
+    # If there's only one distinct top-level exception in the group,
+    # use it and its first-path children, but throw out the exception group and any copies.
+    # For example, Group<['Da', 'Da', 'Da']> should just be treated as a single 'Da'.
+    # We'll also set the main_exception_id, which is used in the extract_metadata function
+    # in src/sentry/eventtypes/error.py - which will ensure the issue is titled by this
+    # item rather than the exception group.
+    if len(distinct_exceptions) == 1:
+        event.data["main_exception_id"] = distinct_exceptions[0].mechanism.exception_id
+        return list(get_first_path(distinct_exceptions[0]))
+
+    # When there's more than one distinct top-level exception, return one of each of them AND the root exception group.
+    # NOTE: This deviates from the original RFC, because finding a common ancestor that shares
+    # one of each top-level exception that is _not_ the root is overly complicated.
+    # Also, it's more likely the stack trace of the root exception will be more meaningful
+    # than one of an inner exception group.
+    distinct_exceptions.append(nodes[0])
+    return distinct_exceptions
+
+
+def get_top_level_exceptions(node: SingleException) -> Generator[SingleException, None, None]:
```

Comment:
That shouldn't happen unless the incoming data is garbage.  But I agree we should be more resilient to failure here.

To prevent that possibility, we'd need to keep track of which nodes have been visited already, and fail gracefully if we're visiting a node that's been visited before.

We could either just return (stop yielding new items at that point), or raise an exception to be caught higher up.

Is there a standard way we throw an exception to indicate an error in processing?

Reply:
we have no way of surfacing this as bad data in the product. I would go for the first option.

Reply:
I've seen a red processing errors dialog box before.  Is that just data to add somewhere to the event?

Reply:
that makes sense to me but you have to find a way to implement this sort of exception handling yourself. I doubt that you want to actually raise an exception since you probably want to still control grouping of bad data, but something like `context.add_error()` might be nice.

Reply:
I wasn't able to figure out how to get the processing error added to the event.  (Though I didn't try super hard. 😅)

However, I did add several bad-data tests to prove that we're not going to blow the stack, including one that has a cycle.